### ISSUE TITLE
[🐛 Bug]: RemoteWebDriver cannot be cast to HasDevTools when using RemoteWebDriver to authenticate

### ISSUE THREAD

**Issue by dj1357** [SeleniumHQ/selenium#10413]
### What happened?

I get this error when using RemoteWebDriver to authenticate:

org.openqa.selenium.remote.RemoteWebDriver$ByteBuddy$p5bjR6n0 cannot be cast to org.openqa.selenium.devtools.HasDevTools

I'm aware of  #9934 but the suggested pom solution does not resolve the problem.

See my code and pom below.

### How can we reproduce the issue?

```shell
import java.io.File;
import java.io.IOException;
import java.net.MalformedURLException;
import org.openqa.selenium.By;
import org.openqa.selenium.HasAuthentication;
import org.openqa.selenium.UsernameAndPassword;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.chrome.ChromeDriverService;
import org.openqa.selenium.chrome.ChromeOptions;
import org.openqa.selenium.devtools.DevTools;
import org.openqa.selenium.devtools.HasDevTools;
import org.openqa.selenium.remote.Augmenter;
import org.openqa.selenium.remote.RemoteWebDriver;

/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
/**
 *
 * @author 
 */
public class ChromeDevToolsAugmenterAuthentication {

    public static void main(String[] args) throws MalformedURLException, InterruptedException, IOException {

        ChromeDriverService service;
        //starts the selenium service

        service = new ChromeDriverService.Builder()
                .usingDriverExecutable(new File("/Users/Test/chromedriver"))
                .usingAnyFreePort()
                .build();
        service.start();

        ChromeOptions chromeOptions = new ChromeOptions();
        WebDriver driver = new RemoteWebDriver(service.getUrl(), chromeOptions);

        Augmenter augmenter = new Augmenter();

        driver = augmenter.augment(driver);

        DevTools devTools = ((HasDevTools) driver).getDevTools();
        devTools.createSession();

        driver = augmenter.
                addDriverAugmentation("chrome", HasAuthentication.class, (caps, exec) -> (whenThisMatches, useTheseCredentials) -> devTools.getDomains().network().addAuthHandler(whenThisMatches, useTheseCredentials)).augment(driver);

        ((HasAuthentication) driver).register(UsernameAndPassword.of("admin", "admin"));
        driver.get("https://the-internet.herokuapp.com/basic_auth");
        System.out.println(driver.findElement(By.tagName("p")).getText());
        driver.quit();

        service.stop();
    }

}






<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <groupId>com.mycompany</groupId>
    <artifactId>Tester</artifactId>
    <version>1.0-SNAPSHOT</version>
    <packaging>jar</packaging>
    <properties>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
        <maven.compiler.source>1.8</maven.compiler.source>
        <maven.compiler.target>1.8</maven.compiler.target>
    </properties>
    <pluginRepositories>
        <pluginRepository>
            <id>central</id>
            <name>Central Repository</name>
            <url>https://repo.maven.apache.org/maven2</url>
            <layout>default</layout>
            <snapshots>
                <enabled>false</enabled>
            </snapshots>
            <releases>
                <updatePolicy>never</updatePolicy>
            </releases>
        </pluginRepository>
    </pluginRepositories>
    <repositories>
        <repository>
            <id>central</id>
            <name>Central Repository</name>
            <url>https://repo.maven.apache.org/maven2</url>
            <layout>default</layout>
            <snapshots>
                <enabled>false</enabled>
            </snapshots>
        </repository>
    </repositories>
    <dependencies>
        <dependency>
            <groupId>org.seleniumhq.selenium</groupId>
            <artifactId>selenium-java</artifactId>
            <version>4.1.2</version>
        </dependency>
    </dependencies>
    <build>
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-shade-plugin</artifactId>
                <version>3.2.4</version>
                <executions>
                    <execution>
                        <phase>package</phase>
                        <goals>
                            <goal>shade</goal>
                        </goals>
                        <configuration>
                            <transformers>
                                <transformer implementation="org.apache.maven.plugins.shade.resource.ManifestResourceTransformer">
                                    <mainClass>ChromeDevToolsAugmenterAuthentication</mainClass>
                                </transformer>
                                <transformer implementation="org.apache.maven.plugins.shade.resource.ServicesResourceTransformer"/>
                            </transformers>
                        </configuration>
                    </execution>
                </executions>
            </plugin>
        </plugins>
    </build>
</project>
```

### Relevant log output

```shell
Starting ChromeDriver 99.0.4844.51 (d537ec02474b5afe23684e7963d538896c63ac77-refs/branch-heads/4844@{#875}) on port 60265
Only local connections are allowed.
Please see https://chromedriver.chromium.org/security-considerations for suggestions on keeping ChromeDriver safe.
ChromeDriver was started successfully.
Mar 06, 2022 11:31:03 AM org.openqa.selenium.remote.tracing.opentelemetry.OpenTelemetryTracer createTracer
INFO: Using OpenTelemetry for tracing
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Mar 06, 2022 11:31:05 AM org.openqa.selenium.remote.ProtocolHandshake createSession
INFO: Detected dialect: W3C
Exception in thread "main" java.lang.ClassCastException: org.openqa.selenium.remote.RemoteWebDriver$ByteBuddy$sgG5HxQ9 cannot be cast to org.openqa.selenium.devtools.HasDevTools
	at ChromeDevToolsAugmenterAuthentication.main(ChromeDevToolsAugmenterAuthentication.java:45)
[ERROR] Command execution failed.
org.apache.commons.exec.ExecuteException: Process exited with an error: 1 (Exit value: 1)
    at org.apache.commons.exec.DefaultExecutor.executeInternal (DefaultExecutor.java:404)
    at org.apache.commons.exec.DefaultExecutor.execute (DefaultExecutor.java:166)
    at org.codehaus.mojo.exec.ExecMojo.executeCommandLine (ExecMojo.java:982)
    at org.codehaus.mojo.exec.ExecMojo.executeCommandLine (ExecMojo.java:929)
    at org.codehaus.mojo.exec.ExecMojo.execute (ExecMojo.java:457)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
```

### Operating System

macOS BigSur

### Selenium version

Java 4.1.2

### What are the browser(s) and version(s) where you see this issue?

Chrome 99

### What are the browser driver(s) and version(s) where you see this issue?

ChromeDriver 99.0.4844.51

### Are you using Selenium Grid?

No

**Comment by pujagani** [comment]
Thank you for sharing the details. The issue points to another issue that provided the pom solution https://github.com/SeleniumHQ/selenium/issues/9803#issuecomment-988747773. Since you mentioned that has not worked for you, I am going to need more details as to how do you run the java code shared above. Do you run it via the IDE or do you create a jar using the maven-shade plugin and then run that jar?

**Comment by dj1357** [comment]
Thank you for your email. I’m running it from Netbeans IDE 12.2. However, I
also get the same error message when running from the command line using
Java -jar.

**Comment by titusfortner** [comment]
The `Augmenter` class is assuming that you are using the `RemoteWebDriver` class with a Selenium server, which will contain the Chrome DevTools Protocol URL in the `se:cdp` capability.

Is there a reason you don't want to use the `ChromeDriver` class?

**Comment by titusfortner** [comment]
@diemol / @pujagani 
Augmenter is already kind of a hack, but I'm not sure the right place to hack onto it. This would work, but maybe there's a better way?

org.openqa.selenium.devtools
DevToolsProvider
```
  private String getCdpUrl(Capabilities caps) {
    Object cdp = caps.getCapability("se:cdp");
    if (cdp instanceof String) {
      return (String) cdp;
    }
    String key;
    switch(caps.getBrowserName()) {
    case "chrome":
      key = "goog:chromeOptions";
      break;
    case "msedge":
      key = "ms:edgeOptions";
      break;
    case "firefox":
      key = "moz:firefoxOptions";
      break;
    default:
      return null;
    }
    return String.valueOf(CdpEndpointFinder.getReportedUri(key, caps));
  }
}
```

**Comment by pujagani** [comment]
Thank you Titus for looking into this and providing a solution. I was able to reproduce the error and it happens only when DriverService is used while creating the RemoteWebDriver instance and if that is augmented. It seems like in the list of matching interfaces HasDevTools is not being picked up and the changes you suggested helps in fixing the `ClassCastException` bit. However, a similar change is needed to create a connection since https://github.com/SeleniumHQ/selenium/blob/702b3aae7385adf331d75cdeacd79720cf6b9c84/java/src/org/openqa/selenium/devtools/SeleniumCdpConnection.java#L60 also relies on "se:cdp". I have made changes for the same locally. Will be adding more tests and commit the changes to help fix this.
